<a href="https://pandas.pydata.org/">
<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/ed/Pandas_logo.svg/2560px-Pandas_logo.svg.png" width="300px">
</a>


# Data Wrangling with pandas



## Objectives
This is the demonstration notebook for the first task. The facilitators will present a lecture that is based around this notebook. Before performing the first task as a team, we recommend you first go through this walkthrough together, as it will show a lot of techniques and code that will be useful in the task.  


The main objective of this walkthrough is to familiarize you with the [pandas python package](https://pandas.pydata.org/). Pandas is a widely-used popular Python package that can discover, clean and structure tabular data. In particular, this notebook will cover:
* The basic concept behind the core data structures within Pandas: [Series](https://pandas.pydata.org/docs/user_guide/dsintro.html#series) and [DataFrames](https://pandas.pydata.org/docs/user_guide/dsintro.html#dataframe)
* How to process, organize and clean data stored in DataFrames
* Visualization of data stored within a DataFrame.

After completing this notebook you will be able to:
* Understand key features of the Pandas library.
* Import tabular data from spreadsheets into Pandas
* Understand your data, its size, features and its structure.
* Handle missing values
* Correct data format
* Transform variables, including grouping continuous values and normalization
* Perform statistical analysis
* Visualise data using Matplotlib library


## What do we mean by data wrangling?
Data wrangling is the process of reading in the raw data and then cleaning and processing it so that the resulting information is in suitable format for analysis. Part of the data wrangling process includes data discovery, data cleaning (identifying and then removing any erroneous data), handling missing data, transforming variables, and visualizing the resulting data.

Data wrangling is often needed before any quantitative analysis can begin; however, it is often the most time-consuming and tedious part of the process. There are many different skills and actions that are required as part of data wrangling, many of which are illustrated in the example code below. Many of these steps are part of the functionality that Pandas provides. 

Here are some common, often used steps that are considered part of data wrangling:

* **Discovering**: 
Reviewing and understanding the input data to better understand its structure and what variables will be useful for your problem 
* **Structuring**:
Standardising the format for disparate types of data and make the data usable for automated or semiautomated data analysis. The data must be structured to fit the analytics model 
COME BACK TO THIS

* **Cleaning** : 
There are often missing or implausible values in a dataset. These occur for a number of reasons and they could adversely affect the reliability and repeatability of an analysis. Cleaning techniques are identify and in some cases correct these instances so that they can be handled appropriately in subsequent analyses.

* **Enriching**
You are pretty familiar with the data at this point. This is the moment to ask yourself if you want to enhance the data. Are you looking to add other data to it?


*  **Validating** 
This step involves iterative programming steps that authenticate your data’s quality and safety. For example, you may have problems if your data is not clean or enriched and the attributes are not distributed evenly.

Many of these different stages of data wrangling require you to visualize the data in some form other than the base tabular data. Visualisation libraries such as matplotlib and seaborn are often used to plot data as well as perform some basic statistical analysis. For example, one simple way of identify outliers for cleaning is to plot the data points and visually identify points that appear to be well outside the plausible range of the rest of the data for a given variable. 



## Table of Contents
* [1. Importing pandas and reading in data](#1-importing-pandas-and-reading-in-data)
* [2. Discovering and reviewing the data](#2-discovering-and-reviewing-the-data)
 
    <li><a href="#Identify_missing_Data">Identify and handle missing values</a></li>     
    <li><a href="#deal_missing_values">Deal with missing values</li>
    <li><a href="#correct_data_format">Correct data format</a></li>
    <li><a href="https://#data_standardization">Data standardization</a></li>
    <li><a href="https://#data_normalization">Data normalization (centering/scaling)</a></li>


## 1. Importing pandas and reading in data
### 1.1 Importing packages
The first thing that we need to do is to import the Python packages we are going to need. This includes pandas, [numpy](https://numpy.org/) (which Pandas uses to store data in the two-dimensional tabular structure), and [matplotlib](https://matplotlib.org/), a popular Python package used for data visualisation.

For convenience, we will import these packages with the widely agreed shortname (`pd`,`np`,`plt`), so that we don't always have to provide the package name each time we use some functionality from the package. 

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt


###  1.2 Reading in data
For this notebook, we are using a dataset provided by the World Health Organisation. There are many sources where you might get data from that you will then want to work with using pandas. The most common is to open a file containing tabular data. The most well-known format for tabular data are Excel spreadsheets. Another commonly used format are plain-text *comma separated values* files with the extension `csv`. Each line of a csv file represents a row, and each column is separated by a comma. 

Besides local files, you may also get data by querying a database or requesting data from the web. 

All of these sources generate tabular data that can be stored within pandas DataFrames.

For the purpose of this notebook, we have assumed you have [downloaded the data](https://liveuclac.sharepoint.com/:x:/r/sites/TeamCodersEventsPlanning/Shared%20Documents/General/WHO/who_case_statistics_modified3.csv?d=wec3e29c1cb9d43779f37cc43bf9eed3b&csf=1&web=1&e=wy7e84) and stored it to a local file in the same directory as this notebook. 

pandas has many different functions that will read in data depending on the data format. Below, we will read the filename and store it the variable `input_file` and then use the [read_csv](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) function to read this file 
and store it into a DataFrame, under a variable named `df_raw`. Here the df stands for DataFrame, so that you know what is stored in the variable.  

In [3]:

# If you have not downloaded the file into this directory
# please ammend the first line so that input_file contains
# the full location (path) of the file
input_file = "who_case_statistics_modified3.csv" 
df_raw = pd.read_csv(input_file)

## 2. Discovering and reviewing the data 
Before we can work with the data, we need to know about more about the contents of the data. Here are some core functions that can help you start to review the structure and contents of your DataFrame. 
* `pandas.DataFrame.head()` - prints out the first *n* rows of the DataFrame 
* `pandas.DataFrame.tail()` - prints out the last *n* rows of the DataFrame

You can understand a bit more of the data structure using the following commands:
* `pandas.DataFrame.info()` - summary information (column names, counts, data types) 
* `pandas.DataFrame.describe()` - provides basic summary statistics for each variable

Some basic low-level information about the DataFrame can be extracted with the following attributes:
* `pandas.DataFrame.shape` - provides the number of rows and columns in the DataFrame
* `pandas.DataFrame.dtypes` - indicates the datatype (string, float, integer, etc) used to store the data for each variable


### 2.1 Review the data
Use the method <b>head()</b> to display the first five rows of the dataframe.


In [11]:
df_raw.head()

country  year     sex          age cases_no  population  HDI for year  \
0  Albania  1987    male  15-24 years       21      312900           NaN   
1  Albania  1987    male  35-54 years       16      308000           NaN   
2  Albania  1987  female  15-24 years       14      289700           NaN   
3  Albania  1987    male    75+ years        1       21800           NaN   
4  Albania  1987    male  25-34 years        9      274300           NaN   

   gdp_for_year ($)   
0      2,156,624,900  
1      2,156,624,900  
2      2,156,624,900  
3      2,156,624,900  
4      2,156,624,900

In [6]:
df_raw.shape

(27841, 8)

In [4]:
df_raw.dtypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27840 entries, 0 to 27839
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   country             27840 non-null  object 
 1   year                27840 non-null  int64  
 2   sex                 27840 non-null  object 
 3   age                 27840 non-null  object 
 4   cases_no            23575 non-null  object 
 5   population          27840 non-null  int64  
 6   HDI for year        8368 non-null   float64
 7    gdp_for_year ($)   27840 non-null  object 
dtypes: float64(1), int64(2), object(5)
memory usage: 1.7+ MB


### Exercise 1
Use the tail method to see the last part of the data.
How many lines do you see?
Use the the tail method again to see the last 20 lines of the data.

In [ ]:
# Write your code here

### Exercise 2
Use the info method to gather information  about the data. 


In [ ]:
# Write your code here

### Observations
* The data frame contains 27841 rows and 8 columns.
* The data frame is an object of class pandas.DataFrame.
* Columns that hold only numeric data are stored as floats or integers
* Columns that hold text or a mixture of data are stored as objects. 


### Exercise 3
Write any other observation that you see about the data. 


### 2.2 Understanding headers
How did pandas know how to name the columns? If you opened up the file in a text editor, you would see that the first row contains these names. When you run `read_csv`, pandas assumes that the column names (or _headers_) are located in the first row. If that is not the case, you can specify which row they can be found by using the `headers` argument. 

However, if there are not any column names in the file, then it is important to tell pandas that, lest you lose actual data. This can be done by setting `headers` to `None`. 


In [ ]:
# Reading in the data - but assuming no headers
df_raw = pd.read_csv(input_file, headers=None)
# Now look at the head, does it look correct?
df_raw.head()

In [ ]:
# Also look at the info, does it look correct?
df_raw.info()

### 2.3 Selecting the columns and getting the values
The way to to select columns is to write the 
`dataframe['column_name']`
Lets show all the countries in the data set.

In [5]:
df_raw = pd.read_csv(input_file)
countries=df_raw['country']
print(countries)

0            Albania
1            Albania
2            Albania
3            Albania
4            Albania
            ...     
27835        Belgium
27836       Thailand
27837    Netherlands
27838        Grenada
27839         Mexico
Name: country, Length: 27840, dtype: object


We can get the counts of different values in the countries column using the `pandas.DataFrame.value_counts()` function.

In [6]:
df_raw['country'].value_counts() #select the column with data_raw['country'] then add the method .value_counts()

Austria                   383
Netherlands               383
Iceland                   382
Mauritius                 382
Belgium                   373
                         ... 
Bosnia and Herzegovina     24
Macau                      12
Cabo Verde                 12
Dominica                   12
Mongolia                   10
Name: country, Length: 101, dtype: int64

This shows us how many times each country is found in the column. If you wanted instead to get a list of all countries found in this column, you can use the `unique` function. The output from this will be a one-dimensional array.  


In [12]:
country_array = df_raw['country'].unique()
country_list = df_raw['country'].unique().astype('str')
country_list

array(['Albania', 'Antigua and Barbuda', 'Argentina', 'Armenia', 'Aruba',
       'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain',
       'Barbados', 'Belarus', 'Belgium', 'Belize',
       'Bosnia and Herzegovina', 'Brazil', 'Bulgaria', 'Cabo Verde',
       'Canada', 'Chile', 'Colombia', 'Costa Rica', 'Croatia', 'Cuba',
       'Cyprus', 'Czech Republic', 'Denmark', 'Dominica', 'Ecuador',
       'El Salvador', 'Estonia', 'Fiji', 'Finland', 'France', 'Georgia',
       'Germany', 'Greece', 'Grenada', 'Guatemala', 'Guyana', 'Hungary',
       'Iceland', 'Ireland', 'Israel', 'Italy', 'Jamaica', 'Japan',
       'Kazakhstan', 'Kiribati', 'Kuwait', 'Kyrgyzstan', 'Latvia',
       'Lithuania', 'Luxembourg', 'Macau', 'Maldives', 'Malta',
       'Mauritius', 'Mexico', 'Mongolia', 'Montenegro', 'Netherlands',
       'New Zealand', 'Nicaragua', 'Norway', 'Oman', 'Panama', 'Paraguay',
       'Philippines', 'Poland', 'Portugal', 'Puerto Rico', 'Qatar',
       'Republic of Korea', 'Romania', '

In [ ]:
type(data_raw_countries_alphabetically)

The data_raw_countries_alphabeticallly is a numpy array that we need to convert it to list, using the **.tolist()** method.


In [ ]:
data_countries_list= data_raw_countries_alphabetically.tolist() # Convert a numpy array to a list
data_countries_list  # The list starts and finishes with []
#  we can print the list within the print function:      print(f'The list of countries is:{data_countries_list}')

In [ ]:
#Find the length of the list
len(data_countries_list)

### Data columns Investigation 
### Find the mean of a numerical column
The method to find the mean of a column is **.mean()**

In [ ]:
data_raw_suicides = data_raw['suicides_no']
data_raw_suicides

In [ ]:
data_raw_suicides_mean = data_raw['suicides_no'].mean()

### ERROR !
 We observe that:
There are  string values at the suicide_no column and we cannot find the mean with the .mean() method because it applies only to integers or strings!

Lets to display all the unique values in the column of 'suicides_no' with the method **.unique()** and change the type of these values to integers using typecasting **.astype(int)**.
 

In [ ]:
data_raw_suicides = data_raw['suicides_no'].unique().astype(int) # Convert the data into the suicides_no column into an integer
data_raw_suicides

### Error and Observations:
   * There are many NaN (Not a Number) data points in the suicide_no column, that indicate missing values and these cannot converted into integer.
   * As displayed at the data_raw above the HDI column is dominated by NaN. We need to delete the column with the Human Development Index because we will not use it for the analysis and it has lots of missing values that we cannot deal with them. We will also save space and time. 
   


# 3) **How to work with missing data?**
As we can see, there are several NaN (not a number) in the data frame, it might also be word  Null or any other words and string values. Those are missing values which may hinder our further analysis. I'm showing here, how do we identify all those missing values and work with missing data in order to bring our Data Frame at a format that is ready for analysis.

Steps for working with missing data:

</ol>
    <li>Identify missing data</li>
    <li>Deal with missing data</li>
    <li>Correct data format</li>
</ol>





<h2 id="Identify_missing_Data">3.1 Identify missing Data</h2>


### **3.1.1 Find missing data**
The missing values are converted by default to Null. We use the following methodsto identify these missing values:

- **.isnull()**
- **.notnull()**
The output is a boolean value indicating whether the value that is passed into the argument is in fact missing data.



When we use the **isnull()** method the output is True when there is Null, and False when there is a real value. The opposite happens when we use the .notnull() method

In [ ]:
(data_raw.isnull()).value_counts() # Indicates with True that there are Nulls and counts the values that are numbers 

There are Null values in just two columns at the suicides_no and at the HDI for year column

The opposite we observe with **.notnull()** Similar to not null is to use the **(~)** in front of the dataset and then the .isnull(). The tilde sign **(~)** indicates negation.  

In [ ]:
(data_raw.notnull()).value_counts() # Indicates with True that we have values and with False where there are Nulls and counts the values that are numbers (not Null)

In [ ]:
(~data_raw.isnull()).value_counts() # The ~ indicates negation and here in front of isnull results to notnull()

#### We can have the same result but reverted Booleans by using the **.notnull()** method.

**(~data_raw.isnull()).value_counts()** # The tilde symbol indicates negation and here in front of isnull results to notnull()

**(data_raw.notnull()).value_counts()** # Indicates with False that there are Nulls and counts the values that are numbers 


In [ ]:
missing_data = data_raw.isnull() # find data that are null and return 'True' if is null and 'False' if it is not
missing_data.head(20)

### **3.1.2.Count the missing values in the data frame**.

### Count missing values in each column

Using a for loop in Python, we can quickly figure out the number of missing values in each column. As mentioned above, "True" represents a missing value and "False" means the value is present in the dataset.  In the body of the for loop the method **".value_counts()"** counts the number of data values, (number of  "False" here). 




In [ ]:
missing_data=(data_raw.isnull())

In [ ]:
type(missing_data)

In [ ]:
# The for loop 
for column in data_raw.columns: # Select each column-name in the header and prints it 
    print(column)
    print(missing_data[column].value_counts())
    print('----')

In [ ]:
#The commented code give the same result as the one that follows but is shortened as puts the methods together.
#for column in missing_data.columns.values.tolist():  # Take the data Frame missing_data,selects the columns and their values and makes a list of them.  
 #print(column)
 #print (missing_data[column].value_counts())
 ## print("") 

An easier way is to show the data set 

In [ ]:
missing_data = data_raw.isnull() # find data that are null and return 'True' if is null and 'False' if it is not
missing_data.head(10)

As discussed the only columns that contain missing data are the "suicides_no'and the HDI.

In [ ]:
print (data_raw.isnull().sum()) # Find the data that are Null in the whole data frame  

The length of the dataframe is 27841 and the column suicides_no has 23575 numbers. It is shorter than the total length of the data 27841 by 27841- 23575= 4265. It is indicated as True (is null) in the count. The missing values therefore there must be strings characters in the column. We need to investigate it.

The HDI column has lots of missing data 19472 values are missing. We will not use it for the analysis and it will be good to delete this column.

## **3.2 Investigation for alpharithmetic characters or words misplaced as values**
There are two ways to identify the non-integer values in the column suicides_no
1) Using the module Regular expressions which we need to import as re. 
A regular expression is a special sequence of characters that helps you match or find other strings or sets of strings, using a specialized syntax held in a pattern.

2) Find the non numeric data with the method **.str.isdigit()** and set it to False that way it collects all the data points that are strings.

### 1st Way Regular expressions module **re**

If you want to run the following code uncomment it from the import downwards.
As I have this notebook run I have selected the 2nd method (.str.isdigit())

In [ ]:
# Import module Regular expressions as re.
#The re module provides an interface to the regular expression engine, 
# allowing you to compile REs into objects and then perform matches with them


#import re 
#replace = re.compile("([a-zA-Z]+)")  # compile any alphabetic character

#data_raw['string'] = data_raw['suicides_no'].str.extract(replace) # Adds a column string to the data set with all the strings which have letters
#data_raw['integer'] = data_raw['suicides_no'].str.replace(replace, " ") #Adds a column integer to the data set 

#data_raw['string'].unique() # Presents the unique "words" from the string column

 In the column suicide-no there are the strings **'Null'**, **'Unknown'** and **nan (not a number)** and we need to replace them.

### 2nd Way without is to use the **.isdigit()** method to a string.


The following code takes all the strings in the column which have digits and sets it to False. Which means that takes all the strings that have no digits. Using the method **unique()** identifies which are these strings.

In [ ]:
non_numeric=data_raw['suicides_no'].str.isdigit() == False

In [ ]:
type (non_numeric)# The non_numeric is 1D array, Series 

In [ ]:
data_raw['suicides_no'][non_numeric].unique()

We observe that there are not only the NaN strings but there are also the "Null" and "Unknown" strings and we need to remove them

We cannot see the NaN values and we apply the pd.numeric (explain)

It's an elegant way to find the strings and words in the datasets without the regular expresions, however it doesn't show us the NaN datapoints. We can add this with the following coding which is more complete and shorten than regular expresions (re).

In [ ]:
suicides =data_raw['suicides_no']

In [ ]:
suicides2 = suicides.apply(pd.to_numeric, errors='coerce')

In [ ]:
print(data_raw['suicides_no'][suicides2.isna()].unique())# Shows the data which are not numeric

The above shows that in the data there are points which as NaN (not an number) and have the word Null and Unknown.

In [ ]:
data_raw.head(15)

In [ ]:

df_null=data_raw.loc[data_raw['suicides_no']=='Null']
df_null

In [ ]:
data_raw.head(15)

#### In the column suicide-no there are the strings 'Null', 'Unknown' and NaN and we need to replace them.

<h2 id="deal_missing_values">4 Deal with missing data</h2>
<b>How to deal with missing data?</b>

<ol>
    <li>Drop data<br>
        a. Drop the whole row<br>
        b. Drop the whole column
    </li>
    <li>Replace data<br>
        a. Replace it by mean<br>
        b. Replace it by frequency<br>
        c. Replace it based on other functions
    </li>
</ol>



### 4.1  Drop / Delete a column with lots of missing data.
To remove a column we use the .drop() method as **df.drop(["column1, column2, ... "], axis=1)**. the axis =1 indicates that I apply the method to columns.
 * I will cut the column HDI for year, There are a lot of missing values there and I don't think that if contributes to the result.
 
 * I will name the new Data Frame **data_frame** because it is not the raw data anymore


In [ ]:
#If you have run the Regular expressions and you have created the columns 'string' and 'integers', please run the code I have commented. 
#data_frame= data_raw.drop(['HDI for year','string','integer'], axis = 1) 

# If  you have run the 2nd way to identify the string values, which is with the .isdigit(), run the following command. 
data_frame= data_raw.drop(['HDI for year'], axis = 1)                         
data_frame

So far we have investigated that there are 4265 missing integers in the suicide_no column which would be the strings 
'NaN', 'Unknown', "Null".

We have deleted the HDI for year colum because it had a lot of missing values

We have created a new data_frame with the data, which is not longer the raw data.

###  4.2 REPLACE :
#### Locate the data in suicides_no that appeared as strings and REPLACE them

In [ ]:
df_null=data_frame.loc[data_frame['suicides_no']=='Null']
df_null

In [ ]:
df_null.shape

There are 4265 points at the suicides_no with the wrong/missing values. It's exactly what we have calculated with the missing_values variable.

In [ ]:
#### Count the values of the 'Null' in 'suicides_no' column 
#data_null["suicides_no"].value_counts()  # Same as:  data_null.suicides_no.value_counts()  
df_null.suicides_no.value_counts()  

### If we want to be concise we could write it elegantly in one line
# data_frame.loc[data_frame['suicides_no']=='Null'].suicides_no.value_counts()

In [ ]:
df_unknown=data_frame.loc[data_frame['suicides_no'] =='Unknown'] 
df_unknown  ## show the lines with the word Unknown in the suicides_no column 

In [ ]:
df_unknown.suicides_no.value_counts() ## count the values with the word Unknown in the suicides_no column 

## REPLACE  incorrect / missing values!
There are three types of data in the column 'suicides_no' in our Data Frame that need to be cleared out. The word 'Null',the word "Unknown" and NaN (Not a Number). I will replace the word Null' with 0 , because its is actually zero. 

The strategy with other words which don't actually contribute to data (unless it is categorical data), is to replace them with the string NaN.  I replace " " with NaN (Not a Number), which is Python's default missing value marker, for reasons of computational speed and convenience. Then we can delete all the NaN data points.
- Replace all strings with NaN
- Delete all NaN values

Here we use the functions to use to localise and replace one value with another:

* .loc[data['Column Name']=='Character to localise']
to localise the Characters in the Column name
* .replace(A, B, inplace = True) 
to replace A value  by B value.



In [ ]:
# replace an empty space " " and "Unknown" to NaN and the "Null" with 0 using Numpy library!!
data_frame.replace(" ", np.nan, inplace = True)
data_frame.replace("Null", 0, inplace = True)
data_frame.replace("Unknown", np.nan, inplace = True)
data_frame.head(15)

### Lets count now how many columns are NaN

In [ ]:
#df_NaN=data_frame.loc[data_frame['suicides_no'] =='NaN'] 
#df_NaN.suicides_no.value_counts()

Based on the summary above, each column has 27840 rows of data and only one of the columns the suicide_no containes missing data:

<ol>
    <li>"suicides_no": 4273 missing data which are replaced with NaN</li>
    
</ol>


### DELETE/DROP  NaN 
In order to deal with NaN we have to create a numpy array where the NaN appears. We can visualise the data frame by calling .head() or .tail()

The way to  delete the NaN is  using the *.dropna()*  to the  data. Immediately afterwards you need to  to reset the index of the data frame as there are deleted values.


In [ ]:
data_frame.tail(10).reset_index()  # use .reset_index()  to visualise it and count it better

### Remove the NaN values 
We can see the the entry points 27824, 27825, 27828  have NaN values at the suicides_no column. We use **.dropna()**. Always reset index 

In [ ]:
# Now I will cut any data points with NaN since I don't know with what to replace them
data_frame = data_frame.dropna() # deletes the rows with NaN
data_frame.tail(10).reset_index()  ## reset the values 

In [ ]:
data_frame.head()

Whole columns or rows should be dropped only if most entries in the column are empty. In our dataset, none of the columns are empty.


There is some freedom in choosing which method to replace data; however, some methods may seem more reasonable than others. We will apply each method according to the data and what the column represents :

**Replace by mean:** depending on the concept the  missing data could be replaced them with mean
   

#### Calculate the mean value for the "suicides_no" column
It does not make any sense with this data but I display how to substitute with the mean to show the method. 
If you want to use it elsewhere you can remove the # in front of the code and use them 


In [ ]:
# How to calculate and show the average of a column
avg_suicide_no=data_frame['suicides_no'].astype('int').mean(axis=0)# First we need to make the values floats (type casting)
print("Average suicides:",avg_suicide_no )

In [ ]:
print ("Average suicides integer:", int(avg_suicide_no))

Replace "NaN" with the integer average value in the "suicide_no" column.
As written above we will not replace by a mean in this example.
I just show the code to use it when appropriate.



In [ ]:
#data_frame['suicides_no'].replace(np.nan, avg_suicide_no, inplace=True)

### Replace by Frequency:
Find the most frequent value ( the mode) of the column and replace the NaN with the most frequent

In [ ]:
data_frame['suicides_no'].value_counts()

There are 1576 different numbers in the column.  We can also use the ".idxmax()" method to calculate the most common value automatically.

In [ ]:
data_frame['suicides_no'].value_counts().idxmax()

We can see that number **'1' is the most common value**.


The replacement procedure is very similar to what we have seen previously:


In [ ]:
#replace the missing 'NaN' values in the suicides_no column  by the most frequent number which is 1 
#data_frame['suicides_no'].replace(np.nan, 1 , inplace=True)

<b>Good!</b> Now, we have a dataset with no missing values.


<h2 id="correct_data_format"> 5 Correct data format </h2>

We are almost there!
The last step in data cleaning is checking and making sure that all data is in the correct format (int, float, text or other)

In Pandas, we use:

**.dtype()** to check the data type
**.astype()**  to change the data type


<h4>Let's list the data types for each column</h4>


As we can see above, some columns are not of the correct data type. Numerical variables should have type 'float' or 'int', and variables with strings such as categories should have type 'object'. For example, 'year', 'age', 'suicides_no, 'population, and 'gdp_for_year ($)' variables are numerical values, so we should expect them to be of the type 'float' or 'int'; however, they are shown as type 'object'. We have to convert data types into a proper format for each column using the **.astype()** method.

<h4>Convert data types to proper format</h4>


In [ ]:
data_frame[["year", "suicides_no", "population"]] = data_frame[["year", "suicides_no", "population"]].astype("int")
#Type casting using .astype("float")   .astype("str")

## The problems of the column 'gdp_for_year '
* The header of the column has a space before the dolar symbol 'gdp_for_year ($)' and we need to replace it with a   name that is coherent and like 'gdp_for_year_usd'
* The format of the numbers in the column should be without the commas.

### 1st Change the header

In [ ]:
headers= data_frame.columns # Select the headers
headers

The column with the GDP has a header that contains empty spaces and the $ character, these will cause problems whenever we mention the header. We need to change it to a title that would be easy to write and avoid mistakes.

In [ ]:
##There are spaces at the name of the column "  gdp_for_year-usd  " and I'm replacing it with a name without spaces "gdp_for_year_usd"
## #Create a list (headers) and replace with the **lamda* method one 
headers = list(map(lambda x: x.replace(' gdp_for_year ($) ', 'gdp_for_year_usd'), headers)) #Create a list and replace with the **lamda* method one value of the list
headers

In [ ]:
data_frame.columns=headers

### 2nd Change the format 

In [ ]:
data_frame.dtypes

In [ ]:
data_frame['gdp_for_year_usd'] # Show the values, which are of the type object

We need to REMOVE the COMMAS ' on the numbers and typecasting them to integers.

In [ ]:
# First replace the string comma ',' empty space''. Then typecasting the column to integers
data_frame['gdp_for_year_usd']= data_frame['gdp_for_year_usd'].str.replace(',', '').astype(int)
data_frame['gdp_for_year_usd']

In [ ]:
#here the gdp_for_year_usd column is integer and I dont need to run this cell but in case it is not we need to do the typecasting and set it to integer
#data_frame['gdp_for_year_usd'].astype("int")

<h4>Let us list the columns after the conversion</h4>


In [ ]:
data_frame.dtypes

In [ ]:
data_frame.head()

### **Wonderful!**
#### Save the data

Now we have finally obtained the cleaned dataset with no missing values with all data in its proper format.
We can save this data set as it is now the 'clean' data set which is ready for processing. 


In [ ]:
data_frame.to_csv(r'./data/who_suicide_statistics_clean_data.csv', index=False)

# **Data Operations**
Data operations can be performed through various built-in methods for faster data processing and analysis. A few methods are:
- Operations using Grouping
- Operations with Sorting
- Operations with Statistics 
- Operations with Functions (Standarisation and Normalisation)




##  Grouping
 Use  **.groupby('..')** Its a function of grouping the data. It involes a combination of **spliting** the object applying a function and **combine** the results **.groupby('column name, value')**

To select the data of a single country from the 101, I'll  group the data frame by 'country' and from that I will select the country of preference by using **.get_group('selection').
Lets select the data for the United Kingdom.

In [ ]:
country_grouped= data_frame.groupby(['country'])
UK_data=country_grouped.get_group('United Kingdom')
UK_data

We make another data frame by selecting the columns that we want to analyse 

In [287]:
UK_data_year_suicides= UK_data[['country','year', 'suicides_no']] # UK_data = country_grouped.get_group('United Kingdom'). If you want you could add another column as well
UK_data_year_suicides

country  year  suicides_no
26476  United Kingdom  1985          264
26477  United Kingdom  1985          915
26478  United Kingdom  1985          128
26479  United Kingdom  1985           62
26480  United Kingdom  1985          678
...               ...   ...          ...
26843  United Kingdom  2015          181
26844  United Kingdom  2015           18
26845  United Kingdom  2015           14
26846  United Kingdom  2015            6
26847  United Kingdom  2015            4

[369 rows x 3 columns]

To find the total suicides of each year, group two columns together and use the  method **.sum()**


In [ ]:
UK_suicides_per_year =UK_data_year_suicides.groupby(['country','year'],as_index=False).sum()
UK_suicides_per_year.head(10)

## Sorting values 
To sort values either according to their size or alphabetical order we use  the method **.sort_values**

In [ ]:
UK_suicides_per_year_sorted=UK_suicides_per_year.sort_values('suicides_no') # The default is ascending order
UK_suicides_per_year_sorted

## Operations with Statistics 
The methods **.max()**, **.min()**,**.mean()**,**.std()** used to find the maximum, minimum, mean and standard deviation at data. The **.describe** function shows all the statistics.

In [ ]:
UK_suicides_per_year['suicides_no'].max() # Select from the data frame, the column that we are interested and apply the .max() method

In [ ]:
# Use the print() function to print the numbers that you calculate with the .max(), .min(). mean() .std() and all the statistical methods
print('The maximum number of cases in the UK is', UK_suicides_per_year['suicides_no'].max())
print('The minimum number of cases in the UK is',UK_data_year_suicides['suicides_no'].min())
print('The standard deviation of cases in the UK is', UK_suicides_per_year['suicides_no'].std())

## Investigate another country's data and create another sub-data 

Let's repeat what we have calculated and presented for the UK for another country. You could choose any country from the list of the initial data frame. I'm choosing Greece here and I repeat the code above:

In [ ]:
## Investigate another data set 
country_grouped= data_frame.groupby(['country'])
Greece_data=country_grouped.get_group('Greece')
Greece_data

In [ ]:
Greece_data_year_suicides= Greece_data[['country','year', 'suicides_no']] # country_grouped.get_group('Greece').
Greece_data_year_suicides

In [ ]:
Greece_suicides_per_year =Greece_data_year_suicides.groupby(['country','year'],as_index=False).sum()
Greece_suicides_per_year.head(10)

In [ ]:
Greece_suicides_per_year_sorted=Greece_suicides_per_year.sort_values('suicides_no') # The default is ascending order
Greece_suicides_per_year_sorted

In [ ]:
print('The maximum number of cases in Greece is', Greece_suicides_per_year['suicides_no'].max())
print('The minimum number of cases in Greece is',Greece_data_year_suicides['suicides_no'].min())
print('The standard deviation of cases in Greece',Greece_suicides_per_year['suicides_no'].std())



## Combine / Merge the data sets 

Another useful operation is to combine together data sets. Pandas is a powerful library and gives a multifaced approach to combining separate datasets. 
With pandas, you can **merge** , **join** and **concatenate** your datasets, allowing you to unify them , understand and interpret them. 

![merge](./fig/fig_merge.jpg)


If you use the **merge() function** , the default is an inner join, which in most of the cases result in a DataFrames which are shorter. With merge(), you also have control over which column(s) to join on. Let’s say that you want to merge both entire datasets, but only on 'year'  since the combination of the two will yield a unique value for each row. To do so, you can use the on parameter:

In [ ]:
UK_suicides_per_year.shape

In [ ]:
Greece_suicides_per_year.shape

In [ ]:
UK_Greece_merged=pd.merge(UK_suicides_per_year, Greece_suicides_per_year,on='year') # Merges the two datasets (one next to the other) and keep common column the 'year' column
UK_Greece_merged

Depending on the datasets you can join on 'left' or on the 'right but in this example it doesn't mean anything: The indication is with the **how** parameter like below: 

In [ ]:
df1_merge=pd.merge(UK_suicides_per_year, Greece_suicides_per_year,how='right', on='year')
df1_merge 

## Use the .join() method : Combine Data on a Column or index
The merge () is a module function (which means that you write the DataFrames as arguments). The .join() is method.  This enables you to specify only one DataFrame, which will join the DataFrame you call .join() on. 

Here I'll create on data frame (Two_countries_combined) which I'll assign it  the Greek results and then I'll join to it the UK DataFrame.

In [ ]:
one_country= Greece_suicides_per_year

In [ ]:


two_countries_combined=one_country.join(UK_suicides_per_year.set_index(["year"]), on=["year"], how= "inner", lsuffix="_x", rsuffix="_y") # It is important to set the index at th common column. The lsuffix sets the first data Frame on the left
two_countries_combined

In [ ]:
two_countries_combined2=one_country.join(UK_suicides_per_year.set_index(["year"]), on=["year"], lsuffix="_left") # In this datasets if you dont use how "inner" it doesnt matter 
two_countries_combined2

## Use the concat() method : Combine data Across Rows or Columns 
Concatenation is a bit different from the merging techniques with merge() function and .join() method where you would specify the merging (inner or outer), and depending on the type of merge you might lose rows and information. 

With the concatenation the  datasets are stitched together along wither the row axis or the column axis.

If you us the following code that concatenates the dataframes (df1 and df2)  with no parameters the default results will look like the picture:

**df_concatenated = pandas.concat([df1, df2])**



![concatenate1](./fig/fig1_concat.jpg)

This example assumes that your column names are the same.  If your column names are different while concatenating along rows (axis 0), then by default the columns will also be added, and NaN values will be filled in as applicable.

![concatenate2](./fig/fig2_concat.jpg)

If you want to accomplish concatenation along columns instead, you’ll use a concat() call like you did above, but you’ll also need to pass the axis parameter with a value of 1 or "columns":

**df_concatenated = pandas.concat([df1, df2], axis='columns")**  or **df_concatenated = pandas.concat([df1, df2], axis=1)**



![concatenate3](./fig/fig3_concat.jpg)

Lets concatenate our datasets and view them 

In [ ]:
df1_concatenated=pd.concat([UK_suicides_per_year, Greece_suicides_per_year])
df1_concatenated

This results to the datasets being one underneath the other. They are concatenated by default on the "row axis" or axis=0. If you look carefully the indexes are different for each data sets (start from 0-30 for the first dataframe and from 0-30 for the second one)

In [ ]:
df2_concatenated=pd.concat([UK_suicides_per_year, Greece_suicides_per_year], axis =1)
df2_concatenated

You see the data frame 1 next to the other.

By default, a concatenation results in a **set union**, where all data is preserved. You’ve seen this with merge() and .join() as an outer join, and you can specify this with the join parameter.



## **Visualise the data**

Where you are investicating your data of prepare to publish your findings, visualisation is an essential tool. There are various libraries that you can use in Python to visualise your data most popular are Matplotlib, and Seaborn. Here we use the Matplotlib library 

In [ ]:
%matplotlib inline  

  **%matplotlib** is a magic function, if you use it just it displays the figure as a pup up. Using with inline, the figure is inside the noteboook 

In [ ]:
df1_merged=pd.merge(UK_suicides_per_year, Greece_suicides_per_year,how='right', on='year') #
df1_merged 



We can change the column names to be more specific or use the .join() method instead which have arguments lsuffix and rsuffix to specify it. 
Lets change the header as we have seen before at paragraph 1. 


In [ ]:
headers=['UK','year','suicides_no_UK','Greece','suicides_no_Greece']
df1_merged.columns=headers
df1_merged.head()

Now we are ready to plot. Firts import the pyplot module of the matplotlib library as plt.
then use the .plot() function at our dataframe. The arguments start with x column and y column/s

In [ ]:

import matplotlib.pyplot as plt
df1_merged.plot(x='year', y=['suicides_no_UK','suicides_no_Greece'])



The default plot is line plot. There are different other kinds of plots and we need to specify them in the arguments these are : bar, area, barh=horizontal bars, hist= histogramms, box, pie= piecharts, line, scatter. 

In [ ]:
df1_merged.plot(kind= 'bar', x='year', y=['suicides_no_UK','suicides_no_Greece'])

## Operations with Functions Statistical Functions
### Example Statistical Functions of **Normalization** and **Standarization** 
Data is usually collected from different sources in different formats.
Normalization and standardization are the processes of transforming data into a common format, allowing the researcher to make the meaningful comparison.

**Normalization** is the process of transforming values of several variables into a similar range. Typical normalizations include scaling the variable from 0 to 1, scaling the variable so the variance is 



### **Example**

To demonstrate normalization, let's say we want to scale the columns , "suicides_no_UK" and "suicides_no_Greece" at df1_merged data frame.
We would like to normalize those variables so their value ranges from 0 to 1
We replace original value by (original value)/(maximum value)

**Step 1)** Create a function of normalization 

In [ ]:
def normalize(number):
    return (number)/number.max()

**Step 2)** Apply the function to the selected column 

In [ ]:
suicides_UK_normalised=normalize(df1_merged['suicides_no_UK'])
suicides_UK_normalised

In [ ]:
suicides_Greece_normalised=normalize(df1_merged['suicides_no_Greece'])
suicides_Greece_normalised

### **Data Standardization**
Data standardization is also a term for a particular type of data normalization where we subtract the mean and divide by the standard deviation. The data is transformed into a common format with  the mean to be 0 and the variable changes from negative  to positive around the mean. 

In [ ]:
def standarize_test(number):
    return (number-number.mean())/number.std()

In [ ]:
standarize_test(UK_suicides_per_year['suicides_no'])

## Visualise and compare
 I want to make a comparison between the normalised values fo both countries and I use again the merged data.

We can add the two columns of normalised to the df1_merged dataset and create another datasets which includes the statistics. Or we can create a new data set.

Lets go with the first option:

In [ ]:
df1_merged['UK_suicides_normalised']= suicides_UK_normalised # Add a new column at the merged dataframe and add the series of normalisation data as we calculated above
df1_merged['Greece_suicides_normalised']= suicides_Greece_normalised
df1_merged.head()

We can now plot the normalisation data and get insightful information comparing the two countries.

In [ ]:
df1_merged.plot(kind= 'bar', x='year', y=['UK_suicides_normalised','Greece_suicides_normalised'])# we can add in the dataframe the plot()function

You see that in general the % of the cases in the UK was higher until 2011. After 2011 the % cases in Greece were higher and it is due to the economic crisis starting fro 2010 and have the greatest impact to the society after 2011. 
With these kind of comparative plots you can visualise and give insightful information about political and socio-economic of different countries in the 101 one presented in this data. 


## Grouping plots
You can group similar plots in a single figure using subplots. The matplotlib.pyplot.figure() which is shortened to plt.figure() creates a space into which we will place all our plots. The parameter figsize tells Python how big to take this space. Each subplot is placed into the figure using its add_subplot method. The add_subplot method takes 3 parameters. The first denotes how many total rows of subplots there are, the second parameter refers to the total number of subplot columns, and the final parameter denotes which subplot your variable is referencing (left-to-right, top-to-bottom). Each subplot is stored in a different variable (axes1, axes2, axes3). Once a subplot is created, the axes can be titled using the ax1.set_title() or set_xlabel() command (or set_ylabel()). Here are our four plots side by side:

In [ ]:
fig=plt.figure(figsize=(12.0, 8.0))


ax1= fig.add_subplot(2,2,1)
#Here we plot bar using the plt.bar() function and it is set at the ax1 which is axes1 subplot variable
plt.title('UK suicides') 
plt.bar(UK_data['year'],
        UK_data['suicides_no'],
        color='blue')


ax2= fig.add_subplot(2,2,2)
ax2.set_title('GDP in billions USD')
ax2.plot(UK_data['year'],UK_data['gdp_for_year_usd']/10**9, color='purple' )


ax3= fig.add_subplot(2,2,3)

plt.title('Greece suicides')
plt.bar(Greece_data['year'],
        Greece_data['suicides_no'],
        color='orange')


ax4= fig.add_subplot(2,2,4)
ax4.set_title('GDP in billions USD')
ax4.plot(Greece_data['year'],Greece_data['gdp_for_year_usd']/10**9, color='green' )


I want to plot the GDP per capita over the years for both countries
I need to calculate it by dividing the two columns and add the result as an extra column.

Lets plot the sum of the suicides over the years for all countries (101)over the years.  

We could write it in one line: From the dataframe, group by the countries column, summarise and sort the values of suicides and plot the suicides_no

In [ ]:


plt.figure(figsize=(16,8))
data_frame.groupby('country').sum().sort_values(by='suicides_no',ascending=False)[['suicides_no']][:101].plot(kind='bar',figsize=(16,8),title='Sum of suicides by country during 1985-2016')
plt.ylabel("Sum of suicides")
plt.xlabel("Country")

In [ ]:
# Here it plots the top 20 countries in suicides 
plt.figure(figsize=(14,6))
data_frame.groupby('country').sum().sort_values(by='suicides_no',ascending=False)[['suicides_no']][:20].plot(kind='bar',figsize=(16,8),color='purple', title='Sum of suicides of the top 20 countries during 30 years 1985-2016')
plt.ylabel("Sum of suicides")




## Exercise 
Calculate the GDP/capita of the countries and visualise the cases according to GDP/capita.



# Convert to GDP/capita

### Thank you for completing this lab!

## Author
##  Mary Tziraki


                                                                                     

## <h3 align="center"> © Health+Bioscience IDEAS 2022. All rights reserved. <h3/>
